File: package/website/src/api/location.ts
import request from '@/utils/request';
import type { Location, Scene, SceneCreate, SceneUpdate, LocationStatistics, TimelineResponse } from '@/types/location';
import type { Photo } from '@/types/album';

export const locationService = {
  async getYears() {
    const data = await request.get<number[]>('/api/locations/years');
    return data.data;
  },

  async searchLocations(q: string) {
    const data = await request.get<{ label: string, value: { province?: string, city?: string, district?: string } }[]>('/api/locations/search', {
      params: { q }
    });
    return data.data;
  },

  async getLocations(level: 'city' | 'province' | 'district' | 'scene' = 'city', skip: number = 0, limit: number = 100, startDate?: string, endDate?: string) {
    const data = await request.get<Location[]>('/api/locations', {
      params: { level, skip, limit, start_date: startDate || undefined, end_date: endDate || undefined }
    });
    return data.data;
  },

  async getStatistics() {
    const data = await request.get<LocationStatistics>('/api/locations/statistics');
    return data.data;
  },

  async getDistribution(level: 'city' | 'province' | 'district' | 'scene' = 'city', startDate?: string, endDate?: string) {
    const data = await request.get<{name: string, count: number, level: string}[]>('/api/locations/distribution', {
      params: { level, start_date: startDate || undefined, end_date: endDate || undefined }
    });
    return data.data;
  },
  
  async getLocationPhotos(name: string, level: 'city' | 'province' | 'district' | 'scene' = 'city', skip: number = 0, limit: number = 50, startDate?: string, endDate?: string) {
    const data = await request.get<Photo[]>(`/api/locations/${name}/photos`, {
      params: { level, skip, limit, start_date: startDate || undefined, end_date: endDate || undefined }
    });
    return data.data;
  },

  async getTimelineNodes(skip: number = 0, limit: number = 100, startDate?: string, endDate?: string, level: string = 'city') {
    const data = await request.get<TimelineResponse>('/api/locations/timeline', {
      params: { skip, limit, start_date: startDate || undefined, end_date: endDate || undefined, level }
    });
    return data.data;
  },

  async getMapMarkers(startDate?: string, endDate?: string) {
    const data = await request.get<{id: string, lat: number, lng: number}[]>('/api/locations/markers', {
      params: { start_date: startDate || undefined, end_date: endDate || undefined }
    });
    return data.data;
  },

  async getScene(id: string) {
    const data = await request.get<Scene>(`/api/locations/scenes/${id}`);
    return data.data;
  },

  async createScene(scene: SceneCreate) {
    const data = await request.post<Scene>('/api/locations/scenes', scene);
    return data.data;
  },

  async updateScene(id: string, scene: SceneUpdate) {
    const data = await request.put<Scene>(`/api/locations/scenes/${id}`, scene);
    return data.data;
  },

  async getScenesList(skip: number = 0, limit: number = 100, startDate?: string, endDate?: string) {
    const data = await request.get<Scene[]>('/api/locations/scenes/list', {
      params: { skip, limit, start_date: startDate || undefined, end_date: endDate || undefined }
    });
    return data.data;
  },

  async deleteScene(id: string) {
    const data = await request.delete(`/api/locations/scenes/${id}`);
    return data.data;
  }
};

File: package/server/app/api/location.py
from fastapi import APIRouter, Depends, Query, Path, HTTPException
from sqlalchemy.orm import Session
from typing import List
from uuid import UUID
from app.dependencies import get_db
from app.schemas import location as schemas
from app.crud import location as crud
from app.schemas import photo as photo_schemas
from app.schemas import scene as scene_schemas
from app.crud import scene as scene_crud
from app.api import deps
from app.db.models import User

router = APIRouter()

@router.get("/search", response_model=List[schemas.LocationSearchItem], summary="搜索位置")
def search_locations(
    q: str = Query(..., min_length=1, description="搜索关键词"),
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    搜索包含关键词的省/市/区。
    """
    return crud.search_locations(db, current_user.id, q)

@router.get("/years", response_model=List[int], summary="获取所有年份")
def get_years(db: Session = Depends(get_db), current_user: User = Depends(deps.get_current_user)):
    """
    获取所有照片的拍摄年份。
    """
    return crud.get_location_years(db, current_user.id)

@router.get("", response_model=List[schemas.Location], summary="获取位置列表")
def get_locations(
    level: str = Query('city', regex='^(city|province|district|scene)$', description="分组级别：city 或 province 或 district 或 scene"),
    start_date: str = Query(None, description="开始日期"),
    end_date: str = Query(None, description="结束日期"),
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取按城市或省份分组的位置列表，包含每个位置的封面照片和照片数量。
    """
    return crud.get_locations(db, current_user.id, level, skip, limit, start_date, end_date)

@router.get("/distribution", response_model=List[schemas.LocationBase], summary="获取位置分布数据")
def get_location_distribution(
    level: str = Query('city', regex='^(city|province|district|scene)$', description="分组级别：city 或 province 或 district 或 scene"),
    start_date: str = Query(None, description="开始日期"),
    end_date: str = Query(None, description="结束日期"),
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取所有位置的分布数据（仅包含名称和数量），用于地图展示。
    """
    return crud.get_location_distribution(db, current_user.id, level, start_date, end_date)

@router.get("/statistics", response_model=schemas.LocationStatistics, summary="获取位置统计数据")
def get_location_statistics(db: Session = Depends(get_db), current_user: User = Depends(deps.get_current_user)):
    """
    获取位置统计数据（省份、城市、区县数量等）。
    """
    return crud.get_location_statistics(db, current_user.id)

from app.schemas.metadata import PhotoDetail

@router.get("/timeline", response_model=schemas.TimelineResponse, summary="获取足迹时间轴照片")
def get_timeline_photos(
    level: str = Query('city', regex='^(city|province|district|scene)$', description="分组级别：city 或 province 或 district 或 scene"),
    skip: int = 0,
    limit: int = 100,
    start_date: str = Query(None, description="开始日期"),
    end_date: str = Query(None, description="结束日期"),
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取足迹时间轴节点（按天和行程聚合），按拍摄时间倒序排列。
    """
    return crud.get_timeline_nodes(db, current_user.id, level, skip, limit, start_date, end_date)

@router.get("/markers", response_model=List[schemas.MapMarker], summary="获取地图标记点")
def get_map_markers(
    start_date: str = Query(None, description="开始日期"),
    end_date: str = Query(None, description="结束日期"),
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取所有包含GPS信息的照片标记点。
    """
    return crud.get_map_markers(db, current_user.id, start_date, end_date)

@router.post("/scenes", response_model=scene_schemas.Scene, summary="创建景区")
def create_scene(
    scene: scene_schemas.SceneCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    创建新的景区，并自动关联范围内的照片。
    """
    return scene_crud.create_scene(db, scene, owner_id=current_user.id)

@router.get("/scenes/list", response_model=List[scene_schemas.Scene], summary="获取所有景区详情")
def get_scenes_list(
    skip: int = 0,
    limit: int = 100,
    start_date: str = Query(None, description="开始日期"),
    end_date: str = Query(None, description="结束日期"),
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取所有景区详细信息（包含多边形坐标）。
    """
    return scene_crud.get_scenes(db, skip, limit, start_date, end_date, owner_id=current_user.id)

@router.get("/scenes/{scene_id}", response_model=scene_schemas.Scene, summary="获取景区详情")
def get_scene_details(
    scene_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取指定景区的详细信息。
    """
    scene = scene_crud.get_scene(db, scene_id, owner_id=current_user.id)
    if not scene:
        raise HTTPException(status_code=404, detail="Scene not found")
    return scene

@router.put("/scenes/{scene_id}", response_model=scene_schemas.Scene, summary="更新景区")
def update_scene(
    scene_id: UUID,
    scene: scene_schemas.SceneUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    更新景区信息。
    """
    try:
        db_scene = scene_crud.update_scene(db, scene_id, scene, owner_id=current_user.id)
        if not db_scene:
            raise HTTPException(status_code=404, detail="Scene not found")
        return db_scene
    except ValueError as e:
        raise HTTPException(status_code=403, detail=str(e))

@router.delete("/scenes/{scene_id}", summary="删除景区")
def delete_scene(
    scene_id: UUID = Path(..., description="景区ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    删除景区。系统默认景区不允许删除。
    """
    try:
        scene = scene_crud.delete_scene(db, scene_id, owner_id=current_user.id)
        if not scene:
            raise HTTPException(status_code=404, detail="Scene not found")
        return {"status": "success", "message": "Scene deleted"}
    except ValueError as e:
        raise HTTPException(status_code=403, detail=str(e))

@router.get("/{name}/photos", response_model=List[PhotoDetail], summary="获取位置照片列表")
def get_location_photos(
    name: str = Path(..., description="位置名称"),
    level: str = Query('city', regex='^(city|province|district|scene)$', description="分组级别：city 或 province 或 district 或 scene"),
    start_date: str = Query(None, description="开始日期"),
    end_date: str = Query(None, description="结束日期"),
    skip: int = 0,
    limit: int = 50,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_user)
):
    """
    获取指定位置（城市或省份）的照片列表。
    """
    return crud.get_location_photos(db, current_user.id, name, level, skip, limit, start_date, end_date)
